File: ios/Sources/CapacitorUwbPlugin/CapacitorUwbPlugin.swift
import Foundation
import Capacitor
import NearbyInteraction

@objc(CapacitorUwbPlugin)
public class CapacitorUwbPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorUwbPlugin"
    public let jsName = "CapacitorUwb"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDiscoveryToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startPeerSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startControllerSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startControleeSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = CapacitorUwb()

    override public func load() {
        super.load()
        implementation.plugin = self
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        let status = implementation.isAvailable()
        call.resolve([
            "supported": status.supported,
            "available": status.available,
            "platform": "ios"
        ])
    }

    @objc func getDiscoveryToken(_ call: CAPPluginCall) {
        do {
            let token = try implementation.getDiscoveryToken()
            call.resolve([
                "discoveryToken": token
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func startPeerSession(_ call: CAPPluginCall) {
        guard let peerDiscoveryToken = call.getString("peerDiscoveryToken") else {
            call.reject("peerDiscoveryToken is required")
            return
        }

        let isCameraAssistanceEnabled = call.getBool("isCameraAssistanceEnabled") ?? false

        do {
            try implementation.startPeerSession(
                peerDiscoveryToken: peerDiscoveryToken,
                isCameraAssistanceEnabled: isCameraAssistanceEnabled
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func startControllerSession(_ call: CAPPluginCall) {
        call.reject("Controller sessions are only available on Android.")
    }

    @objc func startControleeSession(_ call: CAPPluginCall) {
        call.reject("Controlee sessions are only available on Android.")
    }

    @objc func stopSession(_ call: CAPPluginCall) {
        implementation.stopSession()
        call.resolve()
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }

    func notifyRangingUpdate(object: NINearbyObject) {
        var payload: [String: Any] = [
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]

        if let distance = object.distance {
            payload["distanceMeters"] = distance
        }

        if let direction = object.direction {
            payload["direction"] = [
                "x": direction.x,
                "y": direction.y,
                "z": direction.z
            ]
        }
        if #available(iOS 16.0, *) {
            if let horizontalAngle = object.horizontalAngle {
                payload["horizontalAngleRadians"] = horizontalAngle
            }
        }

        notifyListeners("rangingUpdate", data: payload)
    }

    func notifySessionStateChanged(state: String, reason: String? = nil) {
        var payload: [String: Any] = [
            "state": state
        ]

        if let reason = reason {
            payload["reason"] = reason
        }

        notifyListeners("sessionStateChanged", data: payload)
    }
}

File: android/src/main/kotlin/app/capgo/uwb/CapacitorUwbPlugin.kt
package app.capgo.uwb

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "CapacitorUwb")
class CapacitorUwbPlugin : Plugin() {

    private lateinit var implementation: CapacitorUwb

    override fun load() {
        super.load()
        implementation = CapacitorUwb(context) { event, data ->
            notifyListeners(event, data)
        }
    }

    override fun handleOnDestroy() {
        implementation.destroy()
        super.handleOnDestroy()
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        implementation.isAvailable { result ->
            call.resolve(result)
        }
    }

    @PluginMethod
    fun getDiscoveryToken(call: PluginCall) {
        call.reject("Discovery tokens are only available on iOS Nearby Interaction.")
    }

    @PluginMethod
    fun startPeerSession(call: PluginCall) {
        call.reject("Peer sessions are only available on iOS Nearby Interaction.")
    }

    @PluginMethod
    fun startControllerSession(call: PluginCall) {
        val options = call.data ?: JSObject()
        implementation.startControllerSession(
            options,
            onSuccess = { result ->
                call.resolve(result)
            },
            onError = { message ->
                call.reject(message)
            },
        )
    }

    @PluginMethod
    fun startControleeSession(call: PluginCall) {
        val options = call.data ?: JSObject()
        implementation.startControleeSession(
            options,
            onSuccess = {
                call.resolve()
            },
            onError = { message ->
                call.reject(message)
            },
        )
    }

    @PluginMethod
    fun stopSession(call: PluginCall) {
        implementation.stopSession()
        call.resolve()
    }

    @PluginMethod
    fun getPluginVersion(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("version", implementation.getPluginVersion())
            },
        )
    }
}
